homeassistant/components/remootio/cover.py

```diff
@@ -0,0 +1,179 @@
+    def is_closed(self):
+        """Return True when the Remootio controlled garage door or gate is currently closed."""
+        return self._remootio_client.state == State.CLOSED
+
+    async def async_open_cover(self, **kwargs):
+        """Open the Remootio controlled garage door or gate."""
+        await self._remootio_client.trigger_open()
+
+    async def async_close_cover(self, **kwargs):
+        """Close the Remootio controlled garage door or gate."""
+        await self._remootio_client.trigger_close()
+
+
+class RemootioCoverStateChangeListener(Listener[StateChange]):
+    """Listener to be invoked when Remootio controlled garage door or gate changes its state."""
+
+    _owner: RemootioCover
+
+    def __init__(self, owner: RemootioCover) -> None:
+        """Initialize an instance of this class."""
+        super().__init__()
+        self._owner = owner
+
+    async def execute(self, client: RemootioClient, subject: StateChange) -> None:
+        """Execute this listener. Tell Home Assistant that the Remootio controlled garage door or gate has changed its state."""
+        _LOGGER.debug(
+            "Telling Home Assistant that the Remootio controlled garage door or gate has changed its state. RemootioClientState [%s] RemootioCoverEntityId [%s] RemootioCoverUniqueId [%s] RemootioCoverState [%s]",
+            client.state,
+            self._owner.entity_id,
+            self._owner.unique_id,
+            self._owner.state,
+        )
+        self._owner.async_schedule_update_ha_state()
+
+
+class RemootioCoverEventListener(Listener[Event]):
+    """Listener to be invoked on an event sent by the Remmotio device."""
+
+    _owner: RemootioCover
+
+    def __init__(self, owner: RemootioCover) -> None:
+        """Initialize an instance of this class."""
+        super().__init__()
+        self._owner = owner
+
+    async def execute(self, client: RemootioClient, subject: Event) -> None:
+        """Execute this listener. Fire events in Home Assistant based on events sent by the Remootio device.
+
+        Fires the event remootio_left_open in Home Assistant when the Remootio controlled garage door or gate has been left open.
+        As event data the serial number of the Remootio device (also the id of the entity which represents it in Home Assistant) and
+        the name of the entity which represents the Remootio device in Home Assistant will be passed to the fired event.
+        """
+        if subject.type == EventType.LEFT_OPEN:
+            event_type = f"{DOMAIN.lower()}_{subject.type.name.lower()}"
+
+            _LOGGER.debug(
+                "Firing event. EvenType [%s] RemootioCoverEntityId [%s] RemootioCoverUniqueId [%s]",
+                event_type,
+                self._owner.entity_id,
+                self._owner.unique_id,
+            )
+
+            self._owner.hass.bus.async_fire(
+                event_type,
+                {
+                    ATTR_ENTITY_ID: self._owner.entity_id,
```

Comment:
entities are not allowed to fire events. This must be done on integration level: https://developers.home-assistant.io/docs/integration_events/#what-not-to-do